contracts/minter-imago/src/contract.rs

```diff
@@ -0,0 +1,743 @@
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let action = "mint_sender";
+
+    // If there is no active whitelist right now, check public mint
+    // Check if after start_time
+    if is_public_mint(deps.as_ref(), &info)? && (env.block.time < config.start_time) {
+        return Err(ContractError::BeforeMintStartTime {});
+    }
+
+    // Check if already minted max per address limit
+    let mint_count = mint_count(deps.as_ref(), &info)?;
+    if mint_count >= config.per_address_limit {
+        return Err(ContractError::MaxPerAddressLimitExceeded {});
+    }
+
+    _execute_mint(deps, env, info, action, false, None, None)
+}
+
+// Check if a whitelist exists and not ended
+// Sender has to be whitelisted to mint
+fn is_public_mint(deps: Deps, info: &MessageInfo) -> Result<bool, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // If there is no whitelist, there's only a public mint
+    if config.whitelist.is_none() {
+        return Ok(true);
+    }
+
+    let whitelist = config.whitelist.unwrap();
+
+    let wl_config: WhitelistConfigResponse = deps
+        .querier
+        .query_wasm_smart(whitelist.clone(), &WhitelistQueryMsg::Config {})?;
+
+    if !wl_config.is_active {
+        return Ok(true);
+    }
+
+    let res: HasMemberResponse = deps.querier.query_wasm_smart(
+        whitelist,
+        &WhitelistQueryMsg::HasMember {
+            member: info.sender.to_string(),
+        },
+    )?;
+    if !res.has_member {
+        return Err(ContractError::NotWhitelisted {
+            addr: info.sender.to_string(),
+        });
+    }
+
+    // Check wl per address limit
+    let mint_count = mint_count(deps, info)?;
+    if mint_count >= wl_config.per_address_limit {
+        return Err(ContractError::MaxPerAddressLimitExceeded {});
+    }
+
+    Ok(false)
+}
+
+pub fn execute_mint_to(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    recipient: String,
+) -> Result<Response, ContractError> {
+    let recipient = deps.api.addr_validate(&recipient)?;
+    let config = CONFIG.load(deps.storage)?;
+    let action = "mint_to";
+
+    // Check only admin
+    if info.sender != config.admin {
+        return Err(ContractError::Unauthorized(
+            "Sender is not an admin".to_owned(),
+        ));
+    }
+
+    _execute_mint(deps, env, info, action, true, Some(recipient), None)
+}
+
+pub fn execute_mint_for(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    token_id: u32,
+    recipient: String,
+) -> Result<Response, ContractError> {
+    let recipient = deps.api.addr_validate(&recipient)?;
+    let config = CONFIG.load(deps.storage)?;
+    let action = "mint_for";
+
+    // Check only admin
+    if info.sender != config.admin {
+        return Err(ContractError::Unauthorized(
+            "Sender is not an admin".to_owned(),
+        ));
+    }
+
+    _execute_mint(deps, env, info, action, true, Some(recipient), Some(token_id))
+}
```

Comment:
same here since this will allow to mint an specific token id, would this be an issue for the platform?

Reply:
Oh yea, good call, I will remove the token_id parameter

Reply:
if you only need incremental token ids I don't think you need the map and just mint `last_saved_token` + 1 and store the new sequence after right ?

Reply:
This is a good suggestion. For the sake of time, I will make this change in v2 of this contract.